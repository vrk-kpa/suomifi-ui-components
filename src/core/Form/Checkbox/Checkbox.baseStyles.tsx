import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { element, fixInternalMargins, font } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

/* stylelint-disable no-descending-specificity */
const checkedStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--checked {
    & .fi-checkbox_label {
      &::before {
        border-color: ${theme.colors.highlightBase};
        background-color: ${theme.colors.highlightBase};
      }
      & > .fi-checkbox_icon .fi-icon-base-fill {
        fill: ${theme.colors.whiteBase};
      }
    }
    &.fi-checkbox--disabled {
      & .fi-checkbox_label {
        &::before {
          border-color: ${theme.colors.depthLight1};
          background-color: ${theme.colors.depthLight1};
        }
        & > .fi-checkbox_icon .fi-icon-base-fill {
          fill: ${theme.colors.depthLight3};
        }
      }
    }
    &.fi-checkbox--error {
      & .fi-checkbox_label {
        &::before {
          border-color: ${theme.colors.alertBase};
          background-color: ${theme.colors.alertBase};
          border-width: 2px;
        }
        & > .fi-checkbox_icon .fi-icon-base-fill {
          fill: ${theme.colors.whiteBase};
        }
      }
    }
  }
`;

const disabledStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--disabled {
    & .fi-checkbox_label {
      cursor: not-allowed;
      color: ${theme.colors.depthBase};
      &::before {
        background-color: ${theme.colors.depthLight3};
        border-color: ${theme.colors.depthLight1};
        border-width: 1px;
      }
      & > .fi-checkbox_icon .fi-icon-base-fill {
        fill: ${theme.colors.depthLight3};
      }
    }
    & .fi-hint-text {
      color: ${theme.colors.depthBase};
    }
    &.fi-checkbox--large {
      & .fi-checkbox_label::before {
        border-width: 2px;
      }
    }
  }
`;

const errorStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--error {
    & .fi-checkbox_label {
      &::before {
        border-color: ${theme.colors.alertBase};
        border-width: 2px;
      }
      & > .fi-checkbox_icon .fi-icon-base-fill {
        fill: ${theme.colors.alertBase};
      }
    }
  }
`;

const largeVariantStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--large {
    & .fi-checkbox_label {
      padding-left: ${theme.spacing.xxl};
      line-height: 30px;

      &::before {
        content: '';
        position: absolute;
        left: 0px;
        top: 0px;
        box-sizing: border-box;
        height: 30px;
        width: 30px;
        color: ${theme.colors.depthDark3};
        border: 2px solid;
      }
      & .fi-checkbox_icon {
        height: 20px;
        width: 20px;
        left: 5px;
        top: 4px;
      }
    }
    & .fi-hint-text {
      padding-left: ${theme.spacing.xxl};
    }

    &:focus-within {
      & .fi-checkbox_label {
        &::after {
          width: 32px;
          height: 32px;
          top: -1px;
          left: -1px;
        }
      }
    }
  }
`;

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${font(theme)('bodyTextSmall')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  ${fixInternalMargins()}
  
  & .fi-checkbox_label {
    position: relative;
    display: block;
    padding-left: ${theme.spacing.l};
    cursor: pointer;
    line-height: 27px;

    &::before {
      content: '';
      position: absolute;
      left: 0px;
      top: ${theme.spacing.xxs};
      box-sizing: border-box;
      height: 18px;
      width: 18px;
      border: 1px solid ${theme.colors.depthDark3};
      border-radius: ${theme.radiuses.basic};
      background-color: ${theme.colors.whiteBase};
    }
  }

  & .fi-checkbox_icon {
    position: absolute;
    height: 10px;
    width: 10px;
    left: 4px;
    top: 9px;
  }

  &:focus-within {
    & .fi-checkbox_label {
      /* Modified version of theme.focuses.absoluteFocus */
      &::after {
        content: '';
        position: absolute;
        pointer-events: none;
        top: 4px;
        left: -1px;
        border-radius: 2px;
        background-color: transparent;
        border: 0px solid ${theme.colors.whiteBase};
        box-sizing: border-box;
        box-shadow: 0 0 0 2px ${theme.colors.accentSecondary};
        z-index: 9999;
        width: 20px;
        height: 20px;
        outline: 2px solid transparent; /* For high contrast mode */
      }
    }
  }

  & .fi-checkbox_input {
    position: absolute;
    opacity: 0;
    z-index: -9999;
  }

  & .fi-hint-text {
    padding-left: ${theme.spacing.l};
    color: ${theme.colors.depthDark1};
  }

  & .fi-status-text {
    line-height: 18px;
    &.fi-checkbox_statusText--has-content {
      margin-top: ${theme.spacing.xxs};
    }
  }

  ${largeVariantStyles(theme)};
  ${checkedStyles(theme)};
  ${errorStyles(theme)};
  ${disabledStyles(theme)};
`;
