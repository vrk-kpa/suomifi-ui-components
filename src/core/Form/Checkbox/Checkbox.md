```js
import { Checkbox } from './Checkbox';
import { useState } from 'react';

const [checked, setChecked] = useState(false);
<>
  <Checkbox defaultChecked hintText="This is an example hint text">
    Regular checkbox that is checked and has a hint text
  </Checkbox>

  <Checkbox
    hintText="Example hint text"
    status="error"
    statusText="You need to accept the terms and conditions to continue"
  >
    Regular checkbox with a hint text and an error message
  </Checkbox>

  <Checkbox.large>Large checkbox</Checkbox.large>

  <Checkbox.large defaultChecked hintText="Example hint text">
    Checked large checkbox with a hint text
  </Checkbox.large>

  <Checkbox disabled>Disabled checkbox</Checkbox>

  <Checkbox
    checked={checked}
    onClick={({ checkboxState }) => {
      if (window.confirm('Change checkbox state?')) {
        setChecked(checkboxState);
      }
    }}
  >
    Checkbox with controlled checked state
  </Checkbox>
</>;
```
