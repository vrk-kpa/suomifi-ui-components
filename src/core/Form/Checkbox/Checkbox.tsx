import React, { Component, forwardRef, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { InputStatus, StatusTextCommonProps } from '../types';
import { getConditionalAriaProp } from '../../../utils/aria';
import { getLogger } from '../../../utils/log';
import { AutoId } from '../../utils/AutoId/AutoId';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { HtmlLabel, HtmlDiv, HtmlInput, HtmlInputProps } from '../../../reset';
import { StatusText } from '../StatusText/StatusText';
import { HintText } from '../HintText/HintText';
import { CheckboxGroupConsumer } from './CheckboxGroup';
import { baseStyles } from './Checkbox.baseStyles';
import { IconCheck } from 'suomifi-icons';
import { filterDuplicateKeys } from '../../../utils/common/common';

const baseClassName = 'fi-checkbox';

const checkboxClassNames = {
  container: `${baseClassName}_container`,
  input: `${baseClassName}_input`,
  label: `${baseClassName}_label`,
  statusTextHasContent: `${baseClassName}_statusText--has-content`,
  hintText: `${baseClassName}_hintText`,
  disabled: `${baseClassName}--disabled`,
  error: `${baseClassName}--error`,
  checked: `${baseClassName}--checked`,
  large: `${baseClassName}--large`,
  icon: `${baseClassName}_icon`,
};

type CheckboxStatus = Exclude<InputStatus, 'success'>;

export interface CheckboxProps
  extends StatusTextCommonProps,
    MarginProps,
    Omit<HtmlInputProps, 'onClick' | 'value'> {
  /** Controlled checked state  */
  checked?: boolean;
  /** Default status of Checkbox when not using controlled `checked` state
   * @default false
   */
  defaultChecked?: boolean;
  /** CSS class for custom styles */
  className?: string;
  /** Disables the Checkbox. Value won't be included in a form when submitting */
  disabled?: boolean;
  /** Callback fired on Checkbox click */
  onClick?: ({ checkboxState }: { checkboxState: boolean }) => void;
  /**
   * Children will be rendered as Checkbox's label. Use a short and descriptive label.
   */
  children?: ReactNode;
  /**
   * Variant of the Checkbox
   * @default small
   */
  variant?: 'small' | 'large';
  /**
   * `'default'` | `'error'`
   *
   * Status of the component. Error state creates a red border around the Checkbox.
   * Always use a descriptive `statusText` with an error status.
   * @default default
   */
  status?: CheckboxStatus;
  /**
   * Hint text to be displayed under the label
   */
  hintText?: string;
  /**
   * aria-label for the HTML input element.
   * Alternatively you can define an aria-labelledby
   */
  'aria-label'?: string;
  /**
   * aria-labelledby for the HTML input element.
   * Alternatively you can define an aria-label
   */
  'aria-labelledby'?: string;
  /**
   * aria-describedby for the HTML input element,
   */
  'aria-describedby'?: string;
  /**
   * HTML id attribute.
   * If no id is specified, one will be generated automatically
   */
  id?: string;
  /** HTML name attribute */
  name?: string;
  /** HTML value attribute */
  value?: string;
  /** Ref is passed to the underlying input element. Alternative to React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLInputElement>;
}

class BaseCheckbox extends Component<CheckboxProps> {
  state = {
    checkedState: !!this.props.checked || !!this.props.defaultChecked,
  };

  static getDerivedStateFromProps(
    nextProps: CheckboxProps,
    prevState: { checkedState: boolean },
  ) {
    const { checked } = nextProps;
    if (checked !== undefined && checked !== prevState.checkedState) {
      return { checkedState: checked };
    }
    return null;
  }

  handleClick = () => {
    const { onClick, checked } = this.props;
    const { checkedState } = this.state;
    if (checked === undefined) {
      this.setState({ checkedState: !checkedState });
    }
    if (!!onClick) {
      onClick({ checkboxState: !checkedState });
    }
  };

  render() {
    const {
      id,
      className,
      disabled = false,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      'aria-describedby': ariaDescribedBy,
      statusTextAriaLiveMode = 'assertive',
      children,
      checked: dismissChecked,
      defaultChecked: dismissDefaultChecked,
      onClick: dismissOnClick,
      hintText,
      status,
      statusText,
      name,
      value,
      forwardedRef,
      onClick,
      variant,
      style,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    const { checkedState } = this.state;

    if (!children) {
      getLogger().error(
        'Checkbox component should have a label or a child element that acts as one. Add label content or a child element.',
      );
    }

    if (
      ('name' in this.props && (typeof name !== 'string' || name === '')) ||
      ('value' in this.props && (typeof value !== 'string' || value === ''))
    ) {
      getLogger().warn(
        'Name and value props should have non-empty values if provided.',
      );
    }

    const statusTextId = !!statusText ? `${id}-statusText` : undefined;
    const hintTextId = !!hintText ? `${id}-hintText` : undefined;

    return (
      <HtmlDiv
        className={classnames(
          checkboxClassNames.container,
          className,
          baseClassName,
          {
            [checkboxClassNames.error]: status === 'error' && !disabled,
            [checkboxClassNames.checked]: checkedState,
            [checkboxClassNames.large]: variant === 'large',
            [checkboxClassNames.disabled]: !!disabled,
          },
        )}
        style={style}
      >
        <HtmlInput
          type="checkbox"
          disabled={disabled}
          id={id}
          {...getConditionalAriaProp('aria-label', [ariaLabel])}
          {...getConditionalAriaProp('aria-labelledby', [ariaLabelledBy])}
          {...getConditionalAriaProp('aria-describedby', [
            statusTextId,
            hintTextId,
            ariaDescribedBy,
          ])}
          aria-invalid={status === 'error'}
          checked={!!checkedState}
          className={checkboxClassNames.input}
          onChange={this.handleClick}
          name={name}
          forwardedRef={forwardedRef}
          {...(value ? { value } : {})}
          {...passProps}
        />
        <HtmlLabel htmlFor={id} className={checkboxClassNames.label}>
          {!!checkedState && <IconCheck className={checkboxClassNames.icon} />}
          {children}
        </HtmlLabel>
        <HintText id={hintTextId}>{hintText}</HintText>
        <StatusText
          className={classnames({
            [checkboxClassNames.statusTextHasContent]: !!statusText,
          })}
          id={statusTextId}
          status={status}
          disabled={disabled}
          ariaLiveMode={statusTextAriaLiveMode}
        >
          {statusText}
        </StatusText>
      </HtmlDiv>
    );
  }
}

const StyledCheckbox = styled(
  ({
    theme,
    globalMargins,
    ...passProps
  }: CheckboxProps & SuomifiThemeProp & GlobalMarginProps & MarginProps) => (
    <BaseCheckbox {...passProps} />
  ),
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.checkbox,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Checkbox = forwardRef(
  (props: CheckboxProps, ref: React.RefObject<HTMLInputElement>) => {
    const { id: propId, status: propStatus, ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <AutoId id={propId}>
                {(id) => (
                  <CheckboxGroupConsumer>
                    {({ status: groupStatus }) => (
                      <StyledCheckbox
                        theme={suomifiTheme}
                        id={id}
                        forwardedRef={ref}
                        globalMargins={margins}
                        status={!!propStatus ? propStatus : groupStatus}
                        {...passProps}
                      />
                    )}
                  </CheckboxGroupConsumer>
                )}
              </AutoId>
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

Checkbox.displayName = 'Checkbox';
export { Checkbox };
