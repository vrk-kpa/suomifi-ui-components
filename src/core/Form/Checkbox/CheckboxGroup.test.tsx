import React from 'react';
import { render, screen } from '@testing-library/react';
import { Checkbox } from './Checkbox';
import { CheckboxGroup } from './CheckboxGroup';

const CheckboxChilds = [1, 2, 3].map((value) => (
  <Checkbox key={value} id={`test-id-${value}`}>
    {`Label text ${value}`}
  </Checkbox>
));

describe('default, with only required props', () => {
  const DefaultGroup = (
    <CheckboxGroup labelText="Label">{CheckboxChilds}</CheckboxGroup>
  );

  it('should match snapshot', () => {
    const { container } = render(DefaultGroup);
    expect(container).toMatchSnapshot();
  });
});

describe('props', () => {
  describe('className', () => {
    const ClassNameGroup = (
      <CheckboxGroup labelText="Label" className="custom-className">
        {CheckboxChilds}
      </CheckboxGroup>
    );
    it('has the default className', () => {
      const { container } = render(ClassNameGroup);
      expect(container.firstChild).toHaveClass('fi-checkbox-group');
    });
    it('has the given className', () => {
      const { container } = render(ClassNameGroup);
      expect(container.firstChild).toHaveClass('custom-className');
    });
  });

  describe('groupHintText', () => {
    const HintTextGroup = (
      <CheckboxGroup labelText="Label" groupHintText="Example hint text">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has hintText element', () => {
      const { getByText } = render(HintTextGroup);
      const hintText = getByText('Example hint text');
      expect(hintText).toHaveClass('fi-hint-text');
    });
  });

  describe('id', () => {
    const IdGroup = (
      <CheckboxGroup labelText="Label" id="nice-id">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has the given id', () => {
      const { container } = render(IdGroup);
      expect(container.firstChild).toHaveAttribute('id', 'nice-id');
    });
  });

  describe('labelMode', () => {
    const LabelModeGroup = (
      <CheckboxGroup labelText="Label" labelMode="hidden">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has visually hidden label element', () => {
      const { getByText } = render(LabelModeGroup);
      const label = getByText('Label');
      expect(label).toHaveClass('fi-visually-hidden');
    });
  });

  describe('labelText', () => {
    const LabelTextGroup = (
      <CheckboxGroup labelText="Label">{CheckboxChilds}</CheckboxGroup>
    );

    it('has label element', () => {
      const { getByText } = render(LabelTextGroup);
      const label = getByText('Label');
      expect(label).not.toEqual(null);
    });
  });

  describe('optionalText', () => {
    const OptionalTextGroup = (
      <CheckboxGroup labelText="Label" optionalText="optional">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has optional text element', () => {
      const { getByText } = render(OptionalTextGroup);
      const optionalText = getByText('(optional)');
      expect(optionalText).not.toEqual(null);
    });
  });

  describe('groupStatusText', () => {
    const StatusGroup = (
      <CheckboxGroup labelText="Label" groupStatusText="All good!">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('has the given text', () => {
      render(StatusGroup);
      expect(screen.getByText('All good!')).not.toBeNull();
    });
  });

  describe('groupStatus', () => {
    const StatusGroup = (
      <CheckboxGroup labelText="Label" groupStatus="error">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    const StatusGroupOneDifferent = (
      <CheckboxGroup labelText="Label" groupStatus="error">
        <Checkbox key={1} id={`test-id-${1}`}>
          {`Label text ${1}`}
        </Checkbox>
        <Checkbox key={2} id={`test-id-${2}`} status="default">
          {`Label text ${2}`}
        </Checkbox>
        <Checkbox key={3} id={`test-id-${3}`}>
          {`Label text ${3}`}
        </Checkbox>
      </CheckboxGroup>
    );

    it('has Checkboxes with the error statuses', () => {
      render(StatusGroup);
      const fieldset = screen.getByRole('group');
      expect(fieldset.querySelectorAll('.fi-status-text--error')).toHaveLength(
        3,
      );
    });

    it('has two in error state; not overriding individual Checkbox prop', () => {
      render(StatusGroupOneDifferent);
      const fieldset = screen.getByRole('group');
      expect(fieldset.querySelectorAll('.fi-status-text--error')).toHaveLength(
        2,
      );
    });
  });

  describe('margin', () => {
    const Group = (
      <CheckboxGroup data-testid="group" labelText="Test" margin="xs">
        {CheckboxChilds}
      </CheckboxGroup>
    );

    it('should have margin style from margin prop', () => {
      const { getByTestId } = render(Group);
      const div = getByTestId('group');
      expect(div).toHaveStyle('margin: 10px');
    });

    it('should have margin style overridden by style prop', async () => {
      const { container } = render(
        <CheckboxGroup labelText="Test" margin="xs" style={{ margin: 2 }}>
          {CheckboxChilds}
        </CheckboxGroup>,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });
});
