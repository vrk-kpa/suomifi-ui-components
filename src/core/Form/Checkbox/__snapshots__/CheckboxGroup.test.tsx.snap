// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default, with only required props should match snapshot 1`] = `
.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c8 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c8::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c8::before,
.c8::after {
  box-sizing: border-box;
}

.c9 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c9::before,
.c9::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c5 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c5::before,
.c5::after {
  box-sizing: border-box;
}

.c10 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,13%);
  font-size: 14px;
  line-height: 20px;
}

.c10.fi-status-text {
  display: block;
}

.c10.fi-status-text.fi-status-text--error {
  color: hsl(3,59%,43%);
}

.c10.fi-status-text .fi-icon {
  vertical-align: middle;
  -webkit-transform: translateY(-0.1em);
  -ms-transform: translateY(-0.1em);
  transform: translateY(-0.1em);
  margin-right: 5px;
}

.c4.fi-label .fi-label_label-span {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  display: inline-block;
  vertical-align: middle;
  color: hsl(0,0%,13%);
}

.c4.fi-label .fi-label_label-span .fi-label_optional-text {
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c4.fi-label .fi-label_label-span .fi-tooltip {
  margin-left: 6px;
}

.c1 {
  color: hsl(0,0%,13%);
}

.c1 .fi-checkbox-group_legend--with-margin {
  margin-bottom: 10px;
}

.c1 .fi-checkbox-group_legend .fi-hint-text {
  margin-bottom: 0;
}

.c1 .fi-checkbox-group_label--with-margin {
  margin-bottom: 10px;
}

.c1 .fi-checkbox-group_container > .fi-checkbox {
  margin-bottom: 10px;
}

.c1 .fi-checkbox-group_container > .fi-checkbox:last-child {
  margin-bottom: 0;
}

.c1 .fi-checkbox-group_container .fi-checkbox--large {
  margin-top: 5px;
  margin-bottom: 15px;
}

.c7 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c7 .fi-label,
.c7 .fi-hint-text,
.c7 .fi-status-text {
  margin: 0;
}

.c7 .fi-checkbox_label {
  position: relative;
  display: block;
  padding-left: 25px;
  cursor: pointer;
  line-height: 27px;
}

.c7 .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 5px;
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  border: 1px solid hsl(201,7%,58%);
  border-radius: 2px;
  background-color: hsl(0,0%,100%);
}

.c7 .fi-checkbox_icon {
  position: absolute;
  height: 10px;
  width: 10px;
  left: 4px;
  top: 9px;
}

.c7:focus-within .fi-checkbox_label::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: 4px;
  left: -1px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  width: 20px;
  height: 20px;
  outline: 2px solid transparent;
}

.c7 .fi-checkbox_input {
  position: absolute;
  opacity: 0;
  z-index: -9999;
}

.c7 .fi-hint-text {
  padding-left: 25px;
  color: hsl(202,7%,40%);
}

.c7 .fi-status-text {
  line-height: 18px;
}

.c7 .fi-status-text.fi-checkbox_statusText--has-content {
  margin-top: 5px;
}

.c7.fi-checkbox--large .fi-checkbox_label {
  padding-left: 40px;
  line-height: 30px;
}

.c7.fi-checkbox--large .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 0px;
  box-sizing: border-box;
  height: 30px;
  width: 30px;
  color: hsl(201,7%,58%);
  border: 2px solid;
}

.c7.fi-checkbox--large .fi-checkbox_label .fi-checkbox_icon {
  height: 20px;
  width: 20px;
  left: 5px;
  top: 4px;
}

.c7.fi-checkbox--large .fi-hint-text {
  padding-left: 40px;
}

.c7.fi-checkbox--large:focus-within .fi-checkbox_label::after {
  width: 32px;
  height: 32px;
  top: -1px;
  left: -1px;
}

.c7.fi-checkbox--checked .fi-checkbox_label::before {
  border-color: hsl(212,63%,45%);
  background-color: hsl(212,63%,45%);
}

.c7.fi-checkbox--checked .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(0,0%,100%);
}

.c7.fi-checkbox--checked.fi-checkbox--disabled .fi-checkbox_label::before {
  border-color: hsl(202,7%,80%);
  background-color: hsl(202,7%,80%);
}

.c7.fi-checkbox--checked.fi-checkbox--disabled .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(202,7%,97%);
}

.c7.fi-checkbox--checked.fi-checkbox--error .fi-checkbox_label::before {
  border-color: hsl(3,59%,43%);
  background-color: hsl(3,59%,43%);
  border-width: 2px;
}

.c7.fi-checkbox--checked.fi-checkbox--error .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(0,0%,100%);
}

.c7.fi-checkbox--error .fi-checkbox_label::before {
  border-color: hsl(3,59%,43%);
  border-width: 2px;
}

.c7.fi-checkbox--error .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(3,59%,43%);
}

.c7.fi-checkbox--disabled .fi-checkbox_label {
  cursor: not-allowed;
  color: hsl(202,7%,67%);
}

.c7.fi-checkbox--disabled .fi-checkbox_label::before {
  background-color: hsl(202,7%,97%);
  border-color: hsl(202,7%,80%);
  border-width: 1px;
}

.c7.fi-checkbox--disabled .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(202,7%,97%);
}

.c7.fi-checkbox--disabled .fi-hint-text {
  color: hsl(202,7%,67%);
}

.c7.fi-checkbox--disabled.fi-checkbox--large .fi-checkbox_label::before {
  border-width: 2px;
}

<div>
  <div
    class="c0 fi-checkbox-group c1"
    id="1"
  >
    <fieldset
      class="c2"
    >
      <legend
        class="c3 fi-checkbox-group_legend fi-checkbox-group_legend--with-margin"
      >
        <div
          class="c0 c4 fi-label"
        >
          <label
            class="c5 fi-label_label-span"
            for="1"
          >
            Label
          </label>
        </div>
      </legend>
      <div
        class="c6 fi-checkbox-group_container"
      >
        <div
          class="c6 fi-checkbox_container c7 fi-checkbox"
        >
          <input
            aria-invalid="false"
            class="c8 fi-checkbox_input"
            id="test-id-1"
            type="checkbox"
          />
          <label
            class="c9 fi-checkbox_label"
            for="test-id-1"
          >
            Label text 1
          </label>
          <span
            aria-atomic="true"
            aria-live="assertive"
            class="c5 c10 fi-status-text"
          />
        </div>
        <div
          class="c6 fi-checkbox_container c7 fi-checkbox"
        >
          <input
            aria-invalid="false"
            class="c8 fi-checkbox_input"
            id="test-id-2"
            type="checkbox"
          />
          <label
            class="c9 fi-checkbox_label"
            for="test-id-2"
          >
            Label text 2
          </label>
          <span
            aria-atomic="true"
            aria-live="assertive"
            class="c5 c10 fi-status-text"
          />
        </div>
        <div
          class="c6 fi-checkbox_container c7 fi-checkbox"
        >
          <input
            aria-invalid="false"
            class="c8 fi-checkbox_input"
            id="test-id-3"
            type="checkbox"
          />
          <label
            class="c9 fi-checkbox_label"
            for="test-id-3"
          >
            Label text 3
          </label>
          <span
            aria-atomic="true"
            aria-live="assertive"
            class="c5 c10 fi-status-text"
          />
        </div>
      </div>
    </fieldset>
    <span
      aria-atomic="true"
      aria-live="polite"
      class="c5 c10 fi-status-text"
    />
  </div>
</div>
`;
